lib/MusicBrainz/Server/Controller/Label.pm

```diff
@@ -173,4 +176,44 @@ with 'MusicBrainz::Server::Controller::Role::Delete' => {
     edit_type      => $EDIT_LABEL_DELETE,
 };
 
+around edit => sub {
+    my $orig = shift;
+    my ($self, $c) = @_;
+
+    my $label = $c->stash->{label};
+    if ($label->is_special_purpose) {
+        my %props = (
+            label => $label,
+        );
+        $c->stash(
+            component_path => 'label/SpecialPurpose',
+            component_props => \%props,
+            current_view => 'Node',
+        );
+        $c->response->status(HTTP_FORBIDDEN);
+        $c->detach;
+    }
+    else {
+        $self->$orig($c);
+    }
+};
+
+around _validate_merge => sub {
+    my ($orig, $self, $c, $form) = @_;
+    return unless $self->$orig($c, $form);
+    my $target = $form->field('target')->value;
+    my @all = map { $_->value } $form->field('merging')->fields;
+    if (grep { is_special_label($_) && $target != $_ } @all) {
+        $form->field('target')->add_error(l('You cannot merge a special purpose label into another label.'));
```

Comment:
Also need `use MusicBrainz::Server::Translation qw( l );` imported for these errors to work.